New config style and Release algorythm

destiny117 opened the pull request:
most important change:
- the config file now has a "default" section which is applies to all accounts.
- furthermore many configs were grouped for better visibility and more logical structure

this is because the config file really got bloated and #127 did not receive any easy propositions yet, i hope this is good for now and ppl dont have too much of a hassle adapting their local configs

next i have reimplemented the releasing-algorythm from #110 because it feels way more natural for me.

xezero commented on pgoapi/pgoapi.py:
```diff
@@ -627,39 +640,65 @@ def cleanup_pokemon(self, inventory_items=None):
+        elif pokemon.iv_normalized > self.KEEP_IV_OVER \
+                and pokemons_keep_iv_counter < self.MAX_POKEMON_HIGH_IV \
+                and pokemon.cp > (best_pokemon.cp * self.KEEP_IV_WITH_PERCENT_CP / 100):
+            return False, True
```
Is there a better way to write the return values instead of tuples? It's confusing even with comments :|

destiny117 replied:
the thing is we need 3 return states for the MAX_POKEMON_HIGH_IV feature, i can see that it is a bit confusing at first, but you can simply ignore the second return value for all the normal checks and returns. it's only needed for the IV feature

xezero commented on pgoapi/pgoapi.py:
```diff
@@ -627,39 +640,65 @@ def cleanup_pokemon(self, inventory_items=None):
+            return pokemon.cp + pokemon.iv_normalized
+        elif self.SCORE_METHOD == "FANCY":
+            return (pokemon.iv / 100.0 * self.SCORE_WEIGHT_IV) + \
+                   (pokemon.level / (self.player_stats.level+1.5) * self.SCORE_WEIGHT_LVL)
```
I love the fancy formula :) 👍

xezero commented on pgoapi/pgoapi.py:
```diff
@@ -627,39 +640,65 @@ def cleanup_pokemon(self, inventory_items=None):
+            return pokemon.iv_normalized
+        elif self.SCORE_METHOD == "CP*IV":
+            return pokemon.cp * pokemon.iv_normalized
+        elif self.SCORE_METHOD == "CP+IV":
```
I still believe we should remove CP*IV and CP+IV as options since you're literally multiplying/adding a potentially very large number with a number between 0-100. I can't even comprehend what that means and if someone were to use these methods, it could release a lot of pokemon that's not intended.

I think just the "CP", "IV", and "FANCY" methods would be sufficient and cover most cases, and are more intuitive.

destiny117 replied:
i agree i don't see a reason to use them either, but @beeedy has implemented them so maybe he uses one of them. i'm hesitant to remove a feature that is already live, especially when it does not bother the rest of the code

beeedy replied:
I do use CP+IV frequently. What does it "mean"? nothing, really. Why do I use it? I prefer to sort by CP but the little bump from the IV can help the bot to determine which CP pokemon to keep if they are close and one has dramatically different IV. This assumes you are only dealing with pokemon with CP > 500.

xezero replied:
I was just going by your stance of not having features that could accidentally remove a large number of Pokemon if they don't know what they're doing. Using CP+IV for CP > 1000 for instance would essentially nullify the impact of IV as CP at that point overcomes practically any amount of IV. Plus, feature bloat and configuration bloat is not always a good thing, so just because you think a feature is not useless, doesn't mean it's necessary.

destiny117 replied:
i am totally ok with @beedy s reasoning, after all the default values should be fine and if someone is changing the methods he should read up the definition first anyway.
there is no additional config bloat for this, just the description in the readme and those 4 lines of code

xezero replied:
I'm perfectly fine with them being there so long as the description in the readme explained that those choices come with caveats and/or should be used optimally for certain CP ranges, etc.

If a feature is baked into a product, then it means it should be adding value to the user in some way and not just there with no explanation as to why it's there or how/when to use it.

xezero commented:
Going to close #188 when this one gets accepted, as a lot of what was discussed there has been implemented here, plus digging the new config file! 💃

destiny117 commented:
This PR is now ready for merge, if everyone is OK with the new config syntax.

I have also updated the readme to better document it

joychugh commented:
will look into it in some time, pretty big PR :) takes time

destiny117 commented:
allright thank you! :) i'm off for tonight, will have some more time tomorrow, just tell me if i should revert the config structure changes or when i need to rebase

joychugh commented on README.md:
```diff
@@ -31,35 +31,51 @@ optional arguments:
+     * `KEEP_IV_ONLY_WITH_PERCENT_CP` Do not keep high-iv pokemons if their CP is lower than X% of the best pokemon
+     * `MAX_POKEMON_HIGH_IV` Do not keep more than X amount of high-iv pokemons (those with highest CP are preferred by default)
+   * `RELEASE_METHOD` = "DUPLICATES"
+     * The bot seems to have a bad habit of hoarding pokemon. Enabling this feature (disabled by default) will have the bot automatically transfer pokemon that are duplicates. To determine which pokemon to transfer when duplicates exist, the pokemons are compared according to the `SCORE_METHOD` setting.  The bot will transfer the lowest scoring pokemon, maintaining `MIN_SIMILAR_POKEMON` of each t ...
```
can we change this "The bot will collect all pokemon it encounters , thus collecting a ton of bad ones and filling up space" since the bot does what its configured to now after your changes goes through

destiny117 replied:
i took this traight from @beeedy but will change the first sentence to your proposition  now

joychugh commented on pgoapi/pgoapi.py:
```diff
@@ -623,39 +639,65 @@ def cleanup_pokemon(self, inventory_items=None):
-        else:
-            return True
-
-    def pokemon_lvl(self, pokemon):
```
Suggestion, can we move pokemon score to pokemon class? we can pass it the score method inside and when we create it we just set its score at the creation time thus not needing to calculate it each time..

destiny117 replied:
good suggestion, will pass SCORE_METHOD and SCORE_SETTINGS to the pokemon class...

joychugh commented on pgoapi/pgoapi.py:
```diff
@@ -623,39 +639,65 @@ def cleanup_pokemon(self, inventory_items=None):
+                sorted_pokemons = sorted(pokemons, key=self.pokemon_score, reverse=True)
+                pokemons_keep_counter = 0
+                pokemons_keep_iv_counter = 0
+                for pokemon in sorted_pokemons:
```
since we really never throw away the best pokemon, why not just do sorted_pokemon[min_similar_pokemon:] then no need to check it everytime down there. And we can set pokemon_keep_counter = 1 to start with.

xezero replied:
Do you mean set pokemon_keep_counter = self.MIN_SIMILAR_POKEMON?

destiny117 replied:
the reason i did not do it was for the functionality of `MAX_POKEMON_HIGH_IV` because i wanted it to also increase the `pokemons_keep_iv_counter` for those first few pokemons if they have a high iv

joychugh commented on pgoapi/pgoapi.py:
```diff
@@ -623,39 +639,65 @@ def cleanup_pokemon(self, inventory_items=None):
-        # release defined throwaway pokemons  but make sure we have kept at least 1 (dont throw away all of them)
-        elif pokemon.pokemon_id in self.throw_pokemon_ids:
-            return True
+            return False, False
```
I wanted to see how do we maintain a max of KEEP_POKEMON_ID pokemons, if i hit a hotspot containing pikachus like i did in japan i got 45 pikachus by morning, it could have filled up my bag. so the Keep_pokemon_id should also respect MAX_SIMILAR_POKEMON so if its favorite can be handled separately to return false false
ifs keep pokemon and the keep_pokemon count is >= max_similar pokemon, throw it away. or keep a different counter for MAX_FAVORITE_POKEMON set default to 9999

destiny117 replied:
this is a good point, i just implemented it and handle favorites seperately now, will push later

joychugh commented on pgoapi/pgoapi.py:
```diff
@@ -623,39 +639,65 @@ def cleanup_pokemon(self, inventory_items=None):
+                    eligible, keep_iv = self.is_pokemon_eligible_for_transfer(pokemon, sorted_pokemons[0],
+                                                                              pokemons_keep_counter,
+                                                                              pokemons_keep_iv_counter)
+                    if (eligible and pokemons_keep_counter > self.MIN_SIMILAR_POKEMON)\
```
`if (eligible and (pokemon_keep_counter > self.min_similar_pokemon OR pokemon_keep_counter >= max similar_pokemon))` I think you want this.

destiny117 replied:
you are right again, i changed it now to always check MAX_SIMILAR_POKEMON within the is_eligible method

destiny117 commented:
I have now incorporated all suggestions from @joychugh, TYVM!

Furthermore i moved all the stuff inside the defaults section of the config back to the account making the defaults section optional. I figured a "defaults" section and copying values to the accounts section might confuse some ppl new to this.

there were also some minor fixes, working on a new rebase now

destiny117 commented:
This is now rebased and ready for merge if there is no further input
